internal/controller/metrics.go

```diff
@@ -0,0 +1,103 @@
+limitations under the License.
+*/
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	inferencev1alpha1 "github.com/openshift/instaslice-operator/api/v1alpha1"
+	logr "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// updateMetrics - updates UpdateDeployedPodTotalMetrics, UpdateGpuSliceMetrics and UpdateCompatibleProfilesMetrics
+func (r *InstasliceReconciler) updateMetrics(ctx context.Context, instasliceList inferencev1alpha1.InstasliceList) error {
+	log := logr.FromContext(ctx)
+	for _, instaslice := range instasliceList.Items {
+		// Fetch latest Instaslice state before updating metrics
+		updatedInstaslice, err := r.getInstasliceObject(ctx, instaslice.Name, instaslice.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get latest Instaslice object", "instaslice", instaslice.Name)
+			return err
+		}
+		sortedGPUs := sortGPUs(updatedInstaslice)
+		for _, gpuID := range sortedGPUs {
+			for podUuid, allocation := range updatedInstaslice.Status.PodAllocationResults {
+				if allocation.GPUUUID == gpuID {
+					allocRequest := updatedInstaslice.Spec.PodAllocationRequests[podUuid]
+					nodeName := allocation.Nodename
+					namespace := allocRequest.PodRef.Namespace
+					podname := allocRequest.PodRef.Name
+					profile := allocRequest.Profile
+					freeSlots, usedSlots := getFreeAndUsedSlicesCount(updatedInstaslice, gpuID)
+					// update GpuSliceMetrics
+					if err := r.UpdateGpuSliceMetrics(string(nodeName), gpuID, usedSlots, freeSlots); err != nil {
+						return fmt.Errorf("failed to update GPU slice metrics (nodeName: %s, gpuID: %s): %w", nodeName, gpuID, err)
+					}
+					// update DeployedPodTotalMetrics
+					if err := r.UpdateDeployedPodTotalMetrics(string(nodeName), gpuID, namespace, podname, profile, allocation.MigPlacement.Size); err != nil {
+						return fmt.Errorf("failed to update deployed pod metrics (nodeName: %s): %w", nodeName, err)
+					}
+				}
+			}
+		}
+		// update CompatibleProfilesMetrics
+		if err := r.UpdateCompatibleProfilesMetrics(*updatedInstaslice, updatedInstaslice.Name); err != nil {
+			log.Error(err, "Failed to update Compatible Profiles Metrics", "nodeName", updatedInstaslice.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+// updateMetricsAllSlotsFree - If no allocations exist, update metrics with all slots free
+func (r *InstasliceReconciler) updateMetricsAllSlotsFree(ctx context.Context, instasliceList inferencev1alpha1.InstasliceList) error {
+	log := logr.FromContext(ctx)
+	for _, instaslice := range instasliceList.Items {
+		// Fetch latest Instaslice state before updating metrics
+		updatedInstaslice, err := r.getInstasliceObject(ctx, instaslice.Name, instaslice.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get latest Instaslice object", "instaslice", instaslice.Name)
+			return err
+		}
+		sortedGPUs := sortGPUs(updatedInstaslice)
+		for _, gpuID := range sortedGPUs {
+			nodeName := updatedInstaslice.Name
+			if len(updatedInstaslice.Spec.PodAllocationRequests) == 0 {
```

Comment:
Okay, so we use allocation requests in this code but not in the other. I am curious to know why.

Reply:
I haven't changed the other logic, just utilized what was there. But this is the actual metrics logic and the information needed is there in allocation requests